### PR TITLE
[IDE] Use ContextMenu on the Errors and Tasks pad

### PR DESCRIPTION
**Pull request by iainx**
Use the cross platform ContextMenu on for the Error and Tasks pads' popup menu.
Means that
 - An ugly Gtk menu isn't used on Cocoa
 - Fixes refcycles and leaks

Create a ColumnSelectorMenu for automatically generating context menus for
hiding or showing treeview columns

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs [hunk 1]
```diff
@@ -120,5 +123,77 @@ internal void RegisterRenderForFontChanges (CellRendererText renderer)
...
+		}
+
+		string id;
+
+		public ColumnSelectorMenu (TreeView treeview, string restoreID, params string[] columnNames)
+		{
+			id = restoreID;
+
+			int i = 0;
+			foreach (var column in treeview.Columns) {
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs]: Would be nicer here.

```
var columns = treeview.Columns;
for (int i ... ) {

}
```

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs [hunk 2]
```diff
@@ -120,5 +123,77 @@ internal void RegisterRenderForFontChanges (CellRendererText renderer)
...
+			foreach (var c in Items) {
+				var item = c as CheckBoxContextMenuItem;
+				if (item == null) {
+					continue;
+				}
+				builder.AppendFormat ("{0};", item.Checked);
+			}
+			builder.Remove (builder.Length - 1, 1);
+
+			Console.WriteLine ($" Writing {builder} to {id}");
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs]: Debug code.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs [hunk 3]
```diff
@@ -120,5 +123,77 @@ internal void RegisterRenderForFontChanges (CellRendererText renderer)
...
+	{
+		class ColumnSelectorMenuItem : CheckBoxContextMenuItem
+		{
+			TreeViewColumn column;
+
+			public ColumnSelectorMenuItem (TreeViewColumn column, string label) : base (label)
+			{
+				this.column = column;
+
+				Checked = column.Visible;
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs]: This is one case I'm not sure whether it'll leak.

We have:
TVC -> this via event handler
and
this -> TVC via field

I think we need a `WeakReference<TreeViewColumn>` here.

Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs]: An easy way to test this, without spinning up a profiler, is to make a finalizer and see if it's hit.

Comment (iainx) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs]: The menu is displayed in response to an event on the TreeView, so I'm not sure what the event is creating a ref between the TVC and this?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs (hunk 1 above)
Comment (iainx) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs]: Really? That's just personal preference surely?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs [hunk 4]
```diff
@@ -120,5 +123,77 @@ internal void RegisterRenderForFontChanges (CellRendererText renderer)
...
+		public ColumnSelectorMenu (TreeView treeview, string restoreID, params string[] columnNames)
+		{
+			id = restoreID;
+
+			int i = 0;
+			foreach (var column in treeview.Columns) {
+				var name = i < columnNames.Length ? columnNames[i] : null;
+
+				var item = new ColumnSelectorMenuItem (column, string.IsNullOrEmpty(name) ? column.Title : name);
+				item.Checked = treeview.Columns[i].Visible;
```
Comment (iainx) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs]: Line isn't needed...

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs (hunk 3 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs]: Well, ref via field and ref via the `OnItemClicked` instance method. A finalizer check would be enough to verify this, but I'm erring on the side of this is leaking.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs (hunk 1 above)
Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui.Components/PadTreeView.cs]: Well, you're still using an indexer here, it just looks cleaner to be to see that `columnNames[i]` matches `treeview.Columns[i]`.

**Comment by Therzok** [comment]
Pushed leak fix for ContextMenu in this case. The instance method subscription will no longer cause a leak, now that the `ContextMenuItem` is wrapped into a WeakReference, rather than strong reference via EventHandler.Target

**Pull request merged**